### PR TITLE
cpu/msp430: added periph GPIO driver implementation

### PR DESCRIPTION
**Pull request by haukepetersen**
EDIT
~~Based on #3724~~

And here is also the GPIO driver.

EDIT
~~Known issue: When configuring a pin as external interrupt source, the system crashes after 2-3 interrupt events. The behavior is pretty similar to the crashing on UART RX (-> see #3742). My first guess is that there is something wrong with the context switching on the MSP430, but not sure yet... Will keep looking into this.~~

### CODE REVIEW

**Comment by OlegHahm** [comment]
As I wrote in the other PR. I'm currently out of MSP430-JTAGs.

**Comment by haukepetersen** [comment]
Interrupt issue seems to be fixed. External interrupts work now fine on the `msb-430h`.

The reason seemed to be, that the ISR routines were not declared `naked` while using the `__enter_isr` and `__exit_isr` macros...

**Comment by haukepetersen** [comment]
rebased as dependencies were merged.

File: cpu/msp430fxyz/periph/gpio.c [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            return PORT_6;
+        default:
+            return NULL;
+    }
+}
+
+static inline msp_port_isr_t *_isr_port(gpio_t pin)
+{
+    msp_port_t *p = _port(pin);
+    if (p == PORT_1 || p == PORT_2) {
```
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: parens

File: cpu/msp430fxyz/include/periph_cpu.h [hunk 2]
```diff
@@ -26,7 +26,69 @@
...
+ */
+enum {
+    P0 = 0,                 /**< PORT 0 */
+    P1 = 1,                 /**< PORT 1 */
+    P2 = 2,                 /**< PORT 2 */
+    P3 = 3,                 /**< PORT 3 */
+    P4 = 4,                 /**< PORT 4 */
+    P5 = 5,                 /**< PORT 5 */
+    P6 = 6,                 /**< PORT 6 */
+};
```
Comment (OlegHahm) [inline cpu/msp430fxyz/include/periph_cpu.h]: not used?

Comment (haukepetersen) [inline cpu/msp430fxyz/include/periph_cpu.h]: not yet, but they will be. These enum you basically use, when you configure devices on your board - e.g. you say something like `#define CC110X_CS      GPIO(P0,3)`

Comment (haukepetersen) [inline cpu/msp430fxyz/include/periph_cpu.h]: Although for these CPUs a simple `GPIO(0,3)` might work, I would still leave this in so we are consisted throughout all CPUs (where others have something like `PORT_A` or `PA` defined here...)

File: cpu/msp430fxyz/periph/gpio.c [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+ * @brief   Interrupt context for each interrupt line
+ */
+static isr_ctx_t isr_ctx[ISR_NUMOF];
+
+
+static msp_port_t *_port(gpio_t pin)
+{
+    switch (pin >> 8) {
+        case 1:
+            return PORT_1;
```
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: no PORT_0?

**Comment by OlegHahm** [comment]
confirmed working on MSB-430h and Z1.

File: cpu/msp430fxyz/periph/gpio.c (hunk 3 above)
Comment (haukepetersen) [inline cpu/msp430fxyz/periph/gpio.c]: nope, the msp430 starts with 1

**Comment by haukepetersen** [comment]
@OlegHahm, are you good with this PR? Then I will squash and let Travis try his luck...

File: cpu/msp430fxyz/periph/gpio.c (hunk 3 above)
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: What is https://github.com/RIOT-OS/RIOT/pull/3760/files#diff-bedb5324b18a3d4064038b8dcc94f1f9R75 then?

File: cpu/msp430fxyz/periph/gpio.c [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+{
+    int i = bitarithm_lsb(_pin(pin));
+    return (_port(pin) == PORT_1) ? i : (i + 8);
+}
+
+int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+{
+    msp_port_t *port = _port(pin);
+    /* check if port is valid and pull resistor are valid */
+    if (port == NULL || pullup != GPIO_NOPULL) {
```
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: more parens

File: cpu/msp430fxyz/periph/gpio.c [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    return 0;
+}
+
+int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
+                    gpio_cb_t cb, void *arg)
+{
+    msp_port_isr_t *port = _isr_port(pin);
+
+    /* check if port and pullup resistor configuration are valid */
+    if (port == NULL || pullup != GPIO_NOPULL) {
```
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: guess what

File: cpu/msp430fxyz/periph/gpio.c (hunk 3 above)
Comment (haukepetersen) [inline cpu/msp430fxyz/periph/gpio.c]: mistake threre... Will fix

**Comment by haukepetersen** [comment]
addressed comments

File: cpu/msp430fxyz/periph/gpio.c [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    port->IE &= ~(_pin(pin));
+    /* configure as input */
+    gpio_init(pin, GPIO_DIR_IN, GPIO_NOPULL);
+    /* save ISR context */
+    isr_ctx[_ctx(pin)].cb = cb;
+    isr_ctx[_ctx(pin)].arg = arg;
+    /* configure flank */
+    if (flank == GPIO_BOTH) {
+        return -1;
+    }
```
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: Move this test to the to of the function, please.

File: cpu/msp430fxyz/periph/gpio.c [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    else {
+        _port(pin)->OD &= ~(_pin(pin));
+    }
+}
+
+static inline void isr_handler(msp_port_isr_t *port, int ctx)
+{
+    for (int i = 0; i < PINS_PER_PORT; i++) {
+        if (port->IE & (1 << i) && port->IFG & (1 << i)) {
```
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: and once more

File: cpu/msp430fxyz/periph/gpio.c (hunk 6 above)
Comment (haukepetersen) [inline cpu/msp430fxyz/periph/gpio.c]: ok

**Comment by haukepetersen** [comment]
moved test for invalid flank configuration

**Comment by OlegHahm** [comment]
please squash

**Comment by haukepetersen** [comment]
done

File: cpu/msp430fxyz/periph/gpio.c [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    return 0;
+}
+
+int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
+                    gpio_cb_t cb, void *arg)
+{
+    msp_port_isr_t *port = _isr_port(pin);
+
+    /* check if port, pull resistor and flank configuration are valid */
+    if ((port == NULL) || (pullup != GPIO_NOPULL) || (fank == GPIO_BOTH)) {
```
Comment (OlegHahm) [inline cpu/msp430fxyz/periph/gpio.c]: typo: `s/fank/flank`

**Comment by haukepetersen** [comment]
fixed typo and squashed

**Comment by OlegHahm** [comment]
ACK. Merge if Travis is happy.

**Pull request merged**